### PR TITLE
Add the ability to capture http requests for verification.

### PR DESCRIPTION
**Pull request by AdamBJohnsonx**
* Add the ability to get the actual requests that were intercepted from the interceptor
* Rename a few of the methods from intercept in order to reduce developer confusion
* Add the ability to get a raw count of the requests from the interceptor

This has impact on:
* https://github.com/AzureAD/ad-accounts-for-android/pull/1600
* https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1445

The unit tests for these packages (OK, just MSAL) are *expected* to fail to build, since this changes the names
of some of the methods that they might use.

### CODE REVIEW

File: testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java [hunk 1]
```diff
@@ -69,17 +109,28 @@ public static MockHttpClient install() {
...
                 // return the http interceptor
-                return interceptors.get(matcher);
+                final HttpRequestInterceptor httpRequestInterceptor = interceptors.get(matcher);
+                return new HttpRequestInterceptor() {
+                    @Override
+                    public HttpResponse performIntercept(@NonNull HttpClient.HttpMethod httpMethod, @NonNull  URL requestUrl, @NonNull Map<String, String> requestHeaders, @Nullable byte[] requestContent) throws IOException {
+                        if (sSaveRequests.get()) {
+                            sInterceptedRequests.add(new HttpRequest(url, requestHeaders, method.name(), requestContent, null, null));
+                        }
+                        sTotalRequests.incrementAndGet();
```
Comment (AdamBJohnsonx) [inline testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:130]: This is in the wrong spot.

Comment (AdamBJohnsonx) [reply testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:130]:
```suggestion
```

File: testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java [hunk 2]
```diff
@@ -69,17 +109,28 @@ public static MockHttpClient install() {
...
      * @return the http request interceptor configured for the http method and request url
      */
-    public static HttpRequestInterceptor intercept(
+    public static HttpRequestInterceptor getInterceptor (
             @NonNull final HttpMethod method,
             @NonNull final URL url,
             final Map<String, String> requestHeaders,
             final byte[] requestContent
     ) {
+
```
Comment (AdamBJohnsonx) [inline testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:118]:
```suggestion
    // this is also not quite right, but will work given the current usage model, where getInterceptor is always called
    sTotalRequests.incrementAndGet()
```

File: testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java [hunk 3]
```diff
@@ -69,17 +109,28 @@ public static MockHttpClient install() {
...
      */
-    public static HttpRequestInterceptor intercept(
+    public static HttpRequestInterceptor getInterceptor (
             @NonNull final HttpMethod method,
             @NonNull final URL url,
             final Map<String, String> requestHeaders,
             final byte[] requestContent
     ) {
+    // this is also not quite right, but will work given the current usage model, where getInterceptor is always called
+    sTotalRequests.incrementAndGet()
```
Comment (AdamBJohnsonx) [inline testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:119]:
```suggestion
        sTotalRequests.incrementAndGet()
```

Comment (AdamBJohnsonx) [inline testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:119]:
```suggestion
        // this is also not quite right, but will work given the current usage model, where getInterceptor is always called
        sTotalRequests.incrementAndGet()
```

File: testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java [hunk 4]
```diff
@@ -69,17 +109,28 @@ public static MockHttpClient install() {
...
      */
-    public static HttpRequestInterceptor intercept(
+    public static HttpRequestInterceptor getInterceptor (
             @NonNull final HttpMethod method,
             @NonNull final URL url,
             final Map<String, String> requestHeaders,
             final byte[] requestContent
     ) {
+        // this is also not quite right, but will work given the current usage model, where getInterceptor is always called
+        sTotalRequests.incrementAndGet()
```
Comment (AdamBJohnsonx) [inline testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:119]:
```suggestion
        sTotalRequests.incrementAndGet();
```

File: testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java [hunk 5]
```diff
@@ -58,9 +75,32 @@
...
-        return new MockHttpClient();
+        sTotalRequests.set(0);
+        sInterceptedRequests.clear();
+        return new MockHttpClient(false);
+    }
+
+    /**
+     * Installs a mock http client instance to use in providing the request interceptors.
+     * <p>
+     * Invoke the uninstall() method to remove all the set interceptors
```
Comment (shahzaibj) [inline testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:86]:
```suggestion
     * Invoke the uninstall() method to remove all the set interceptors.
```

File: testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java [hunk 6]
```diff
@@ -48,7 +53,19 @@
      */
     private static final Map<HttpRequestMatcher, HttpRequestInterceptor> interceptors = new HashMap<>();
 
+    private static final AtomicBoolean sSaveRequests = new AtomicBoolean(false);
 
+    private static final AtomicLong sTotalRequests = new AtomicLong(0);
```
Comment (shahzaibj) [inline testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:58]: Should these be member variables instead? At least they are being set per instance

Comment (AdamBJohnsonx) [reply testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:58]: They're being set in a static method, so... no.  This whole class more-or-less works based on it not being used in parallel across multiple tests.

Comment (AdamBJohnsonx) [reply testutils/src/main/java/com/microsoft/identity/internal/testutils/MockHttpClient.java:58]: I could move the setters out to the static "factory" methods...

**Review by iambmelt — APPROVED** [review]
Nothing from me

**Review by shahzaibj — APPROVED** [review]

**Pull request merged**